### ISSUE TITLE
Problem with persistent PVC and erlang cookie

### ISSUE THREAD

**Issue by WyriHaximus** [groundhog2k/helm-charts#718]
Initially, the pod starts just fine, but after updating the Helm deployment with some unrelated changes, or just killing a random pod the following error comes up:

```
PostStartHookError: command '/bin/bash -ec until rabbitmqctl cluster_status >/dev/null; do echo "Waiting for cluster readiness..." 
sleep 5 done rabbitmq-queues rebalance "all" ' exited with 137: 17:04:03.329 [error] Cookie file /var/lib/rabbitmq/.erlang.cookie must 
be accessible by owner only 17:04:09.732 [error] Cookie file /var/lib/rabbitmq/.erlang.cookie must be accessible by owner only 
17:04:09.739 [error] Cookie file /var/lib/rabbitmq/.erlang.cookie must be accessible by owner only 17:04:15.329 [error] Cookie file 
/var/lib/rabbitmq/.erlang.cookie must be accessible by owner only 17:04:15.339 [error] Cookie file /var/lib/rabbitmq/.erlang.cookie must 
be accessible by owner only 17:04:21.034 [error] Cookie file /var/lib/rabbitmq/.erlang.cookie must be accessible by owner only 
17:04:21.040 [error] Cookie file /var/lib/rabbitmq/.erlang.cookie must be accessible by owner only 17:04:27.039 [error] Cookie file 
/var/lib/rabbitmq/.erlang.cookie must be accessible by owner only 17:04:27.045 [error] Cookie file /var/lib/rabbitmq/.erlang.cookie must
 be accessible by owner only 17:04:33.238 [error] Cookie file /var/lib/rabbitmq/.erlang.cookie must be accessible by owner only 
17:04:33.244 [error] Cookie file /var/lib/rabbitmq/.erlang.cookie must be accessible by owner only 17:04:37.337 [error] Cookie file 
/var/lib/rabbitmq/.erlang.cookie must be accessible by owner only 17:04:37.345 [error] Cookie file /var/lib/rabbitmq/.erlang.cookie must
 be accessible by owner only
```

`values.yaml`:
```yaml
replicaCount: 3
storage:
  requestedSize: 5Gi
  className: longhorn
ingress:
  enabled: true
  hosts:
    - host: rabbitmq.home.wyrihaxim.us
      paths:
        - /
k8sPeerDiscoveryPlugin:
  enabled: true
  k8sPeerDiscoveryPlugin: hostname
managementPlugin:
  enabled: true
prometheusPlugin:
  enabled: true
podAnnotations:
  prometheus.io/scrape: "true"
  prometheus.io/port: "15692"
podDisruptionBudget:
  minAvailable: 2
resources:
  limits:
    cpu: 500m
    memory: 512Mi
  requests:
    cpu: 500m
    memory: 512Mi
affinity:
  podAntiAffinity:
    requiredDuringSchedulingIgnoredDuringExecution:
      - topologyKey: "kubernetes.io/hostname"
        labelSelector:
          matchExpressions:
            - key: "app.kubernetes.io/instance"
              operator: "In"
              values:
                - "rabbitmq"
```

**Comment by WyriHaximus** [comment]
Just been looking at https://github.com/groundhog2k/helm-charts/blob/master/charts/rabbitmq/templates/scripts.yaml#L23 moving that line to after the conditional should fix it, right? (Let me try that.)

**Comment by WyriHaximus** [comment]
Was able to confirm that also applying that chmod even if the file already exists resolves that error.

**Comment by WyriHaximus** [comment]
Created #719 to solve this

**Comment by groundhog2k** [comment]
Hi! I understand the fix, but still wondering why it is necessary. Because /var/lib/rabbitmq is a mounted storage path and during first start the erlang cookie file is created in /var/lib/rabbitmq/.erlang.cookie and permissions are set to `chmod 600`. When an update/restart of a rabbitmq pod happens, the /var/lib/rabbitmq/.erlang.cookie file should still exist on the mounted storage and permissions shouldn't have changed.
So when you need to set the permissions for every pod restart, that would mean something has changed the .erlang_cookie file permissions.
As I see you used the "longhorn" storage provider.
The chart was tested on Kubernetes with Docker for Windows and in production usage with a real nfs-storage provider. In both cases I can't reproduce the behavior.
I can only guess that the behavior you see is somehow connected with longhorn storage provider. It is interesting and I need a bit more time to setup a longhorn storage provider to test it.

**Comment by groundhog2k** [comment]
I think the behavior could be related to the following issue description: https://github.com/longhorn/longhorn/issues/1165
This one is also interesting: https://github.com/longhorn/longhorn/issues/1991

You could try the following: Start a pod (like nginx) without root permissions (securityContext!) and having a PVC (longhorn) attached to it. Exec a bash in the containers pod, create a simple file on the mounted volume path and change permissions with chmod. Now kill the pod and wait until K8s restarted it. Then exec a bash again in the container and check if file exists and also thh file permissions.
It would be interesting to know what you found out.

**Comment by groundhog2k** [comment]
I have one more idea... The chart has a default `fsGroup: 999` (see [values.yaml](https://github.com/groundhog2k/helm-charts/blob/master/charts/rabbitmq/values.yaml)). Perhaps longhorn changes the permissions on the filesystem to group 999 (the RabbitMQ user) with every pod restart. That should not change the `rwx` group flags for the file - only the group itself. That could be a longhorn bug or some configuration.

**Comment by groundhog2k** [comment]
Hi! I managed to setup longhorn storage provider on a K8s test cluster and I can confirm that I was able to reproduce the described behavior based on a simple nginx pod with an attached PVC.
As suspected the fsGroup setting is the root cause for all of this. The nfs-storage provider behaves differently then the longhorn provider. For longhorn it 's like described in: https://kubernetes.io/docs/tasks/configure-pod-container/security-context/#configure-volume-permission-and-ownership-change-policy-for-pods
The only thing that is not 100% clear to me, whether it is correct by longhorn to additional set rwx-flags for the group during every mount.

Anyway, a simple fsGroupChangePolicy: "OnRootMismatch" in the podSecurityContext should fix it.
If this helps in your case, then I'm thinking of adapting this in all of the charts here. I guess other charts could be affected by this behavior too. (Especially gitea and nextcloud).

**Comment by WyriHaximus** [comment]
> Anyway, a simple fsGroupChangePolicy: "OnRootMismatch" in the podSecurityContext should fix it. If this helps in your case, then I'm thinking of adapting this in all of the charts here. I guess other charts could be affected by this behavior too. (Especially gitea and nextcloud).

That works flawlessly! Thank you for the quick and thorough response. Could I help you out by filing PR's for `RabbitMQ` and others turning this on by default as a sort of repayment for the time you took to look into this and find a solution?

P.S. Didn't mention this but I'm running `k3s` (on `k3os`) because it's lighter. So I was slightly worried that this `fsGroupChangePolicy: "OnRootMismatch"` might not be in there (yet). Turns out it is and works perfectly.

**Comment by groundhog2k** [comment]
Thank you for your feedback and to offer your help :-)

I think it doesn't matter if you use k3s or another way to build a K8s cluster - at the end it's "just Kubernetes". The important thing is the Kubernetes version. As described in https://kubernetes.io/docs/tasks/configure-pod-container/security-context/ the `fsGroupChangePolicy` is available with Kubernetes 1.20. So I was curious and deployed a simple pod on a "old" Kubernetes 1.19 with longhorn storage provider ....and suprise: I was not able to set the `fsGroupChangePolicy` and got a "unknown field" error during deployment.

I then removed fsGroup completely from the chart.
The group for the file(s) was not 999 anymore, but longhorn still had the behavior of setting `rw` permissions when a pod was restarted - but now for the other group.

Summarized: The solution only works for Kubernetes >=1.20 :-(
Now I'm not sure if it is a good idea to add `fsGroupChangePolicy` by default, because some charts wont work anymore with Kubernetes version <1.20. An if-case with a version check in the helm chart would fix it, but wouldn't solve the problem for older Kubernetes + longhorn storage provider. :-(

I need to rethink all options and maybe I will open a bug ticker for longhorn again, because the overall behavior makes some charts unusable with older Kubernetes versions. (Not only the ones in this repo, but also other ones from Bitnami etc.)

I would like to come back to your offer, when we have found a good solution.

**Comment by joshimoo** [comment]
@groundhog2k the permission changes are actually handled by kubelet outside of longhorn, only starting in v1.23 is it even possible to request kubernetes to delegate that action to the csi driver. More info here: https://kubernetes-csi.github.io/docs/support-fsgroup.html#overview-1

The most likely reason you didn't catch this on your test system (NFS) is that permission changes are disabled by default for unsupported fs (NFS, glusterFSS) as well as most RWX storage. 
```
For certain multi-writer volume types, such as NFS or Gluster, the cluster doesn’t perform recursive permission changes even if the pod has a fsGroup
```

More info here about the recursive chown done by kubelet, this can be really bad for large volumes with lots of files.
https://kubernetes.io/blog/2020/12/14/kubernetes-release-1.20-fsgroupchangepolicy-fsgrouppolicy/#allow-users-to-skip-recursive-permission-changes-on-mount

Info on how to switch to `RootMissmatch`, this was added after the fact, since it could take up to 30 minutes for a volume to become available if it was large and contained lots of files :)
https://kubernetes.io/docs/tasks/configure-pod-container/security-context/#configure-volume-permission-and-ownership-change-policy-for-pods

**Comment by joshimoo** [comment]
TLDR: this isn't a longhorn issue, the reason your observed behavior is differently between your NFS and longhorn test system is that kubelet does not do recursive chowns on NFS even if a FSGroup is specified.

**Comment by groundhog2k** [comment]
Thank you @joshimoo - I read the article (https://kubernetes.io/blog/2020/12/14/kubernetes-release-1.20-fsgroupchangepolicy-fsgrouppolicy/#allow-users-to-skip-recursive-permission-changes-on-mount) again and now it's clear to me.

That means I can't simply fix it in the helm chart itself for all possible Kubernetes versions. Perhaps it's a combination of setting 'fsGroupChangePolicy' for K8s version >=1.20 and additonally doing a `chmod ` during container init phase if it is necessary for some of the applications containers, like in the case above with RabbitMQ.

**Comment by groundhog2k** [comment]
@WyriHaximus : I would like to inform you, that I will take your first solution with chmod during init into the RabbitMQ chart and also for the Graylog chart. That should fix it for the longhorn storage provider in case of an older K8s cluster (<1.20) when 'fsGroupChangePolicy' is not available. For newer Kubernetes versions it should not harm anything.
